File: src/scrapers/strategies/indeed.ts
import { Job, ScraperConfig, ScraperResult } from '../types';
import { HttpScraper } from './httpScraper';

/**
 * Indeed.com job scraper
 * Uses HTTP requests with Cheerio for HTML parsing
 */
export class IndeedScraper extends HttpScraper {
  private readonly baseUrl = 'https://indeed.com';

  constructor(config: ScraperConfig) {
    super(config, 'IndeedScraper');
  }

  /**
   * Scrape job listings from Indeed.com
   * @param config - Scraper configuration
   * @returns Promise resolving to scraper result with jobs
   */
  async scrape(config: ScraperConfig): Promise<ScraperResult> {
    try {
      const url = this.buildUrl(config);
      this.logger.info(`Scraping Indeed with query: ${config.query}`);

      const html = await this.fetchHtml(url);
      const $ = this.parseHtml(html);
      const jobs: Job[] = [];

      // Indeed job card selector
      const jobCards = $('.jobsearch-List-item');

      this.logger.info(`Found ${jobCards.length} job cards`);

      jobCards.each((_, element) => {
        try {
          const title = $(element).find('h2 a span').first().text().trim();
          const company = $(element).find('[data-testid="company-name"]').text().trim();
          const location = $(element).find('[data-testid="text-location"]').text().trim();
          const linkElem = $(element).find('h2 a');
          const link = linkElem.attr('href') || '';
          const description = $(element).find('.job-snippet').first().text().trim();

          // Normalize the URL
          const fullLink = link ? this.normalizeUrl(link, this.baseUrl) : '';

          const job: Partial<Job> = {
            title,
            company,
            location,
            link: fullLink,
            description,
            source: 'indeed',
          };

          if (this.validateJob(job)) {
            jobs.push(this.createJob(job));
          }
        } catch (error) {
          this.logger.warn(`Error parsing job card: ${error instanceof Error ? error.message : 'Unknown'}`);
        }
      });

      this.logger.info(`Successfully extracted ${jobs.length} jobs from Indeed`);

      return {
        success: true,
        data: jobs,
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      this.logger.error(`Indeed scraping failed: ${errorMessage}`);
      return {
        success: false,
        error: errorMessage,
      };
    }
  }

  /**
   * Build Indeed search URL
   * @param config - Scraper configuration
   * @returns Encoded search URL
   */
  private buildUrl(config: ScraperConfig): string {
    const query = encodeURIComponent(config.query || '');
    return `${this.baseUrl}/jobs?q=${query}&sort=date`;
  }
}

File: scrapers/indeed.py
from scrapers.shared.base import ScraplingBaseScraper, run_scraper_cli
from scrapers.shared.models import Job


class IndeedScraper(ScraplingBaseScraper):
    name = "indeed"

    def build_url(self, query: str) -> str:
        base = (self.config.extra or {}).get("base_url", "https://www.indeed.com")
        return f"{base}/jobs?q={query}&sort=date&fromage=3"

    def _wait_selector(self) -> str | None:
        return None

    def _fetcher_options(self) -> dict:
        return {"network_idle": False, "wait": 3000}

    def parse_page(self, page) -> list[Job]:
        jobs: list[Job] = []
        # Indeed uses multiple card selectors depending on A/B testing
        card_selectors = [
            "[class*='job_seen_beacon']",
            ".cardOutline",
            "[data-testid='job-card']",
            "li[class*='job']",
            "div[class*='jobsearch'] > div > div",
        ]
        job_elements = []
        for sel in card_selectors:
            found = page.css(sel)
            if found and len(found) > 0:
                job_elements = found
                self._log("info", f"Found {len(job_elements)} cards with selector: {sel}")
                break

        if not job_elements:
            # Fallback: try finding any element with job-related class
            job_elements = page.css("[class*='card']")

        self._log("info", f"Total card elements: {len(job_elements)}")

        for el in job_elements:
            job = self._parse_card(el)
            if job:
                jobs.append(job)

        return jobs

    def _parse_card(self, el) -> Job | None:
        try:
            # Title - multiple selector attempts
            title = ""
            for sel in [
                "h2 a[data-testid='job-card-title']",
                "a[data-testid='job-card-title']",
                "a.jobCardLink",
                "h2 a span",
                "a[class*='jobtitle']",
                "a[class*='title']",
                "h2 span",
            ]:
                found = el.css(sel)
                if found and found[0].text:
                    title = found[0].text.strip()
                    break

            if not title:
                return None

            # Company
            company = ""
            for sel in [
                "[data-testid='job-card-company-name']",
                "[data-testid='company-name']",
                "span[class*='company']",
                "div[class*='company']",
                "a[class*='company']",
                "[class*='companyName']",
            ]:
                found = el.css(sel)
                if found and found[0].text:
                    company = found[0].text.strip()
                    break

            # Location
            location = ""
            for sel in [
                "[data-testid='job-card-location']",
                "[data-testid='text-location']",
                "div[class*='location']",
                "[class*='location']",
            ]:
                found = el.css(sel)
                if found and found[0].text:
                    location = found[0].text.strip()
                    break

            # Link
            link = ""
            for sel in [
                "h2 a",
                "a.jobCardLink",
                "a[class*='title']",
                "a[href*='/rc/clk']",
                "a[href*='/jobs?']",
                "a[href*='indeed.com']",
            ]:
                found = el.css(sel)
                if found:
                    link = found[0].attrib.get("href", "")
                    break

            if link and not link.startswith("http"):
                base = (self.config.extra or {}).get("base_url", "https://www.indeed.com")
                prefix = "" if link.startswith("/") else "/"
                link = f"{base}{prefix}{link}"

            return Job(
                id=self._generate_id(link),
                title=title,
                company=company,
                location=location,
                link=link,
                description="",
                source="indeed",
            )
        except Exception as exc:
            self._log("warning", f"Error parsing card: {exc}")
            return None


if __name__ == "__main__":
    run_scraper_cli(IndeedScraper)
